File: Programmers/가장가까운글자.py
from collections import defaultdict
"""
s를 순회하면서, s[i]부터 s[0]까지 거꾸로 순회하며 같은 문자를 찾으면, i-j를 idx로 설정하고 break
만약에 시간복잡도가 N**2 밑이여야한다면,
dic을 만들고, s를 순회하면서, dic[s[i]]에 대해서 dic[s[i]][j] == i에 대해 
dic[s[i]][j-1]를 추가. 근데 이게 시간복잡도가 얼마나 더 낮아질지는 모르겠다.
"""
def solution(s):
    answer = []
    
    for i in range(len(s)):
        idx = -1
        for j in range(i-1,-1,-1):
            if s[i]==s[j]:
                idx = i-j
                break
        answer.append(idx)
    
    
    return answer

File: Programmers/가장가까운글자.java
import java.util.*;
/*
각 문자열에서 두번 이상 등장하는 문자에 대해서 같은 두 문자의 거리를 answer[idx]에 저장하여
answer를 return하는 문제.
*/
class Solution {
    public int[] solution(String s) {
        int[] answer = new int[s.length()];
        
        HashMap<Character,Integer> dic = new HashMap<>();
        
        for (int i=0;i<s.length();i++){
            if (!dic.containsKey(s.charAt(i))){
                answer[i]=-1;
                dic.put(s.charAt(i),i);
            }
            else{
                answer[i] = i - dic.get(s.charAt(i));
                dic.put(s.charAt(i),i);
            }
        }
        return answer;
    }
}
